src/adapter/src/coord/timestamp_selection.rs

```diff
@@ -258,6 +261,148 @@ pub trait TimestampProvider {
+        // stay queryable at the chosen timestamp.
+        let read_holds = self.acquire_read_holds(id_bundle);
+        let timeline = Self::get_timeline(timeline_context);
+
+        // To be filled by strong session serializable logic, should it be engaged.
+        let mut session_oracle_read_ts = None;
+
+        let since = self.least_valid_read(&read_holds);
+        let upper = self.least_valid_write(id_bundle);
+        let largest_not_in_advance_of_upper = Coordinator::largest_not_in_advance_of_upper(&upper);
+
+        // We start by establishing the hard constraints that must be applied to timestamp determination.
+        // These constraints are derived from the input arguments, and properties of the collections involved.
+        // TODO: Many of the constraints are expressed obliquely, and could be made more direct.
+        let constraints = {
+            // Constraints we will populate through a sequence of opinions.
+            let mut constraints = Constraints::default();
+
+            // First, we have validity constraints from the `id_bundle` argument which indicates
+            // which collections we are reading from.
+            // TODO: Refine the detail about which identifiers are binding and which are not.
+            let since = self.least_valid_read(&read_holds);
+            let storage = id_bundle.storage_ids.iter().cloned().collect();
+            constraints
+                .lower
+                .push((since.clone(), Reason::StorageInput(storage)));
+            let compute = id_bundle
+                .compute_ids
+                .iter()
+                .flat_map(|(key, ids)| ids.iter().map(|id| (*key, *id)))
+                .collect();
+            constraints
+                .lower
+                .push((since.clone(), Reason::ComputeInput(compute)));
+
+            // The query's `when` may indicates a specific timestamp we must advance to, or a specific value we must use.
+            if let Some(timestamp) = when.advance_to_timestamp() {
+                let catalog_state = self.catalog_state();
+                let ts = Coordinator::evaluate_when(catalog_state, timestamp, session)?;
+                constraints
+                    .lower
+                    .push((Antichain::from_elem(ts), Reason::QueryAsOf));
+                // If the query is at a specific timestamp, we must introduce an upper bound as well.
+                if let QueryWhen::AtTimestamp(_) = when {
+                    constraints
+                        .upper
+                        .push((Antichain::from_elem(ts), Reason::QueryAsOf));
+                }
+            }
+
+            // The specification of an `oracle_read_ts` may indicates that we must advance to it,
+            // except in one isolation mode, or if `when` does not indicate that we should.
+            // At the moment, only `QueryWhen::FreshestTableWrite` indicates that we should.
```

Comment:
This comment, plus Joe's response, seems to indicate that instead of matching on whether or not there is an `oracle_read_ts`, we should first check the isolation level and possible `QueryWhen` as well?

Then if either of those indicate we need to advance to the `oracle_read_ts` we should assert that one was provided.

Not a blocking comment but seems like a good change to make now?